src/malli/core.cljc

```diff
@@ -1961,9 +2102,34 @@
            AST
            (-to-ast [this _] (-to-value-ast this))
            Schema
-           (-validator [_]
-             (let [validator (-memoize (fn [] (-validator (rf))))]
-               (fn [x] ((validator) x))))
+           (-validator [this]
+             (let [ref-validators *ref-validators*
+                   id (-identify-ref-schema this)]
+               (if lazy
+                 (let [rv (ref-validators id)
+                       vol (or rv (volatile! nil))]
+                   (c/assert (not @vol) "invariant: we tie the knot on the way back up")
+                   (if-some [f @vol]
+                     @f
+                     (fn [x]
```

Comment:
Ok, so in the lazy case, the knot-tying is delayed until the validator is run. That makes sense. Could the code for the lazy and eager cases be made to look even more symmetric, so that the added `fn` wrapper stands out as the only difference?